### PR TITLE
NIFI-12643 Added support for FileResourceService in PutGCSObject

### PR DESCRIPTION
**Pull request by balazsgerner**
# Summary

Added support to upload content directly from local file's inputstream (by-passing the content repository) inside the PutGCSObject processor. The default behaviour is upload from flowfile content to be backward compatible. This can be overridden if the `RESOURCE_TRANSFER_SOURCE` property value is set to `FILE_RESOURCE_SERVICE`, and a `FileResourceService` controller service is registered.

[NIFI-12643](https://issues.apache.org/jira/browse/NIFI-12643)

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [x] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [x] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 21

### Licensing

- [ ] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [ ] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [ ] Documentation formatting appears as expected in rendered files

### CODE REVIEW

File: nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/storage/PutGCSObject.java [hunk 1]
```diff
@@ -291,6 +295,8 @@ public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
         descriptors.add(BUCKET);
         descriptors.add(KEY);
         descriptors.add(CONTENT_TYPE);
+        descriptors.add(RESOURCE_TRANSFER_SOURCE);
+        descriptors.add(FILE_RESOURCE_SERVICE);
         descriptors.add(CRC32C);
         descriptors.add(ACL);
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/storage/PutGCSObject.java:301]: I would move the new properties up, before `CONTENT_TYPE`. In that way, the properties would be "grouped" like:
- target (Bucket, Key)
- source
- metadata (Content Type, CRC, etc)
```suggestion
        descriptors.add(BUCKET);
        descriptors.add(KEY);
        descriptors.add(RESOURCE_TRANSFER_SOURCE);
        descriptors.add(FILE_RESOURCE_SERVICE);
        descriptors.add(CONTENT_TYPE);
        descriptors.add(CRC32C);
        descriptors.add(ACL);
```

**Review by turcsanyip — CHANGES_REQUESTED** [review]
Thanks @balazsgerner for your first contribution!
Tested the new functionality and it works as expected, the code also looks good.
Added one minor comment regarding the position of the new properties on the UI.

File: nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/storage/PutGCSObject.java (hunk 1 above)
Comment (balazsgerner) [reply nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/storage/PutGCSObject.java]: thanks, committed the suggestion.

File: nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/storage/PutGCSObject.java [hunk 2]
```diff
@@ -322,199 +328,196 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
...
+                    .getValue();
+            final String key = context.getProperty(KEY)
+                    .evaluateAttributeExpressions(flowFile)
+                    .getValue();
+            final boolean overwrite = context.getProperty(OVERWRITE).asBoolean();
+
+            final FlowFile ff = flowFile;
+            final String ffFilename = ff.getAttributes().get(CoreAttributes.FILENAME.key());
+            final Map<String, String> attributes = new HashMap<>();
+            final ResourceTransferSource resourceTransferSource = ResourceTransferSource.valueOf(context.getProperty(RESOURCE_TRANSFER_SOURCE).getValue());
```
Comment (turcsanyip) [inline nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/storage/PutGCSObject.java:345]: Please use the recently added `PropertyValue.asAllowableValue()` method:
```suggestion
            final ResourceTransferSource resourceTransferSource = context.getProperty(RESOURCE_TRANSFER_SOURCE).asAllowableValue(ResourceTransferSource.class);
```

**Review by turcsanyip — APPROVED** [review]
Thanks @balazsgerner!

+1 LGTM
Merging...

**Pull request closed without merging**